File: src/Asm.BankPlus.Models/Account.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Asm.BankPlus.Models
{
    public class Account
    {
        public Guid Id { get; set; }

        public string Name { get; set; }

        public string Description { get; set; }

        public DateTimeOffset BalanceDate { get; set; }

        public decimal CurrentBalance { get; set; }

        public decimal AvailableBalance { get; set; }

        public bool IncludeInPosition { get; set; }

        public AccountType AccountType { get; set; }

        public AccountController Controller { get; set; }

        public int? ImporterTypeId { get; set; }

        public IEnumerable<VirtualAccount> VirtualAccounts { get; set; }

        public decimal VirtualAccountRemainingBalance
        {
            get => CurrentBalance - (VirtualAccounts?.Sum(v => v.Balance) ?? 0);
        }
    }
}

File: src/Asm.BankPlus.Web/MooBankApp/src/models/Account.ts
import { VirtualAccount } from "./VirtualAccount";

export enum AccountType {
    None = 0,
    Transaction = 1,
    Savings = 2,
    Credit = 3,
    Mortgage = 4,
    Superannuation = 5,
}

export enum AccountController {
    Manual = 0,
    Virtual = 1,
    Import = 2,
}

export type accountId = string;

export interface AccountBase {
    id: accountId;
    name: string;
    description?: string;
}

export interface Account extends AccountBase {
    currentBalance: number;
    availableBalance: number;
    balanceUpdated: Date;
    accountType: AccountType;
    controller: AccountController;
    includeInPosition: boolean;
    importerTypeId?: number;
    virtualAccountRemainingBalance?: number;
    virtualAccounts: VirtualAccount[];
}

export interface ImportAccount {
    importerTypeId: number;
}
